internal/dag/gatewayapi_processor.go

```diff
@@ -366,6 +401,60 @@ func (p *GatewayAPIProcessor) validGatewayTLS(listenerTLS gatewayapi_v1alpha2.Ga
 	return listenerSecret
 }
 
+type crossNamespaceFrom struct {
+	group     string
+	kind      string
+	namespace string
+}
+
+type crossNamespaceTo struct {
+	group     string
+	kind      string
+	namespace string
+	name      string
+}
+
+func (p *GatewayAPIProcessor) validCrossNamespaceRef(from crossNamespaceFrom, to crossNamespaceTo) bool {
+	for _, referencePolicy := range p.source.referencepolicies {
+		// The ReferencePolicy must be defined in the namespace of
+		// the "to" (the referent).
+		if referencePolicy.Namespace != to.namespace {
+			continue
+		}
+
+		// Check if the ReferencePolicy has a matching "from".
+		var fromAllowed bool
+		for _, refPolicyFrom := range referencePolicy.Spec.From {
+			if string(refPolicyFrom.Namespace) == from.namespace && string(refPolicyFrom.Group) == from.group && string(refPolicyFrom.Kind) == from.kind {
+				fromAllowed = true
+				break
+			}
+		}
+		if !fromAllowed {
+			continue
+		}
+
+		// Check if the ReferencePolicy has a matching "to".
+		var toAllowed bool
+		for _, refPolicyTo := range referencePolicy.Spec.To {
+			if string(refPolicyTo.Group) == to.group && string(refPolicyTo.Kind) == to.kind && (refPolicyTo.Name == nil || *refPolicyTo.Name == "" || string(*refPolicyTo.Name) == to.name) {
```

Comment:
is it valid for someone to put `core` as the group for a secret rather than empty? only edge case I can think of since we're checking for equality of the group

Reply:
Good question -- I have another change lined up to drop "core" as a magic string for the core group from the rest of the Gateway API implementation, since it's not mentioned in the Gateway API spec anywhere, and everything indicates "" should be used to refer to the core API group. Decided not to add support for it here and then immediately drop it.